_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,443 @@
+
+![Image of reference-based messaging flow](/assets/images/posts/2021-05-06-broker-tuning-reference-messaging.png)
+
+Referenced-based messaging means more for message passing requires more trips, so end-to-end latency will increase. One major drawback of this approach is that there is no automatic clean up of the data in the external system when the Kafka message is cleaned up.
+
+One way to mitigate the limitations of reference-based messaging is to take a hybrid approach. Send large messages to the data store and process standard-sized messages directly.
+
+#### Inline messaging
+Inline messaging is a complex process that splits messages into chunks that use the same key, which are then combined on output using a stream processor like Kafka Streams.
+
+The basic steps are:
+
+- The client application serializes then chunks the data if the message is too big.
+- The producer uses the Kafka `ByteArraySerializer` or similar to serialize each chunk again before sending it.
+- The consumer receives the chunks, which are assembled before deserialization.
+- The consumer tracks messages and buffers chunks until it has a complete message.
+- Complete messages are delivered in order according to the offset of the first or last chunk for each set of chunked messages.
+- Successful delivery of the complete message is checked against offset metadata to avoid duplicates during a rebalance.
+
+**Inline messaging flow**
+
+![Image of inline messaging flow](/assets/images/posts/2021-05-06-broker-tuning-inline-messaging.png)
+
+Inline messaging does not depend on external systems like reference-based messaging. But it does have a performance overhead on the consumer side because of the buffering required, particularly when handling a series of large messages in parallel. The chunks of large messages can become interleaved, so that it is not always possible to commit when all the chunks of a message have been consumed if the chunks of another large message in the buffer are incomplete. For this reason, buffering is usually supported by persisting message chunks or by implementing commit logic.
+
+#### Configuration to handle larger messages**
+
+You increase message limits by configuring `message.max.bytes` to set the maximum record batch size. You can set the limit at the topic level or the broker level. If you set limit at the broker level, larger messages are allowed for all topics and messages greater than the maximum limit are rejected. The buffer size for the producers (`max.request.size`) and consumers (`message.max.bytes`) must be able to accommodate the larger messages.
+
+### Controlling the log flush of message data
+
+Usually, log flush thresholds are not usually set and the operating system performs a background flush using its default settings. But if you are using application flush management, setting lower flush thresholds might be appropriate if you are looking at ways to decrease latency or you are using faster disks.
```

Comment:
Flushing isn't going to decrease latency.

Reply:
Removed